lib/smart_proxy_container_gateway/container_gateway_api.rb

```diff
@@ -100,18 +104,40 @@ class Api < ::Sinatra::Base
         response.headers['Docker-Distribution-API-Version'] = 'registry/2.0'
 
         unless auth_header.present? && auth_header.basic_auth?
-          one_year = (60 * 60 * 24 * 365)
-          return { token: AuthorizationHeader::UNAUTHORIZED_TOKEN, issued_at: Time.now.iso8601,
-                   expires_at: (Time.now + one_year).iso8601 }.to_json
+          return { token: AuthorizationHeader::UNAUTHORIZED_TOKEN, issued_at: Time.now.rfc3339,
+                   expires_in: 1.year.seconds.to_i }.to_json
         end
 
         token_response = ForemanApi.new.fetch_token(auth_header.raw_header, request.params)
         if token_response.code.to_i != 200
           halt token_response.code.to_i, token_response.body
         else
+          # This returned token should follow OAuth2 spec. We need some minor conversion
+          # to store the token with the expires_at time (using rfc3339).
           token_response_body = JSON.parse(token_response.body)
-          ContainerGateway.insert_token(request.params['account'], token_response_body['token'],
-                                        token_response_body['expires_at'])
+
+          if token_response_body['token'].nil?
+            halt 502, "Recieved malformed token response"
+          end
+
+          token_issued_at = token_response_body["issued_at"]
+          if token_issued_at.nil? || token_issued_at.to_time.nil?
+            halt 502, "Recieved malformed token response"
+          end
+          issued_at = token_response_body["issued_at"].to_time
```

Comment:
This looks redundant. Why both `token_issued_at` and `issued_at`?

The spec says `issued_at` is optional. Quoting the spec:

> issued_at
>    (Optional) The [RFC3339](https://www.ietf.org/rfc/rfc3339.txt)-serialized UTC standard time at which a given token was issued. If issued_at is omitted, the expiration is from when the token exchange completed.

I think in this case it's the time it's the server response time. So perhaps take the time that the server sent?
```suggestion
          token_issued_at = token_response_body.fetch("issued_at", token_response["Date"])&.to_time
          unless token_issued_at
            halt 502, "Received malformed token response"
          end
```

Reply:
Yeah, would be better to leverage `fetch` like this. Same for the other comment, condenses the code a bit and makes it more readable.

So a "Date" field can always be expected ?

Reply:
Not 100% sure. But `token_response["Date"]` would then return `nil`. `nil&.to_time` is `nil`, which is the only way to trigger the `halt 502` below it. If you can count on the `Date` to be present, the `halt 502` code can be dropped. HTTP 500 can be acceptable in that case.

I also assumed `to_time` can parse [RFC2616 dates](https://www.rfc-editor.org/rfc/rfc2616#section-3.3).

Reply:
Looks like to_time has no issue with the RFC2616 dates. Thank you for the good ideas with this fix. Should be fixed up now.